// @netlify/sdk/dist/integration/connect/utils/data.js
export function lmdbStoredDocumentToStackbit(storedDocument, model) {
    const { ...fields } = storedDocument;
    return {
        id: fields._objectId.value,
        type: `document`,
        modelName: model.getOriginalName(),
        context: {},
        updatedAt: fields._updatedAt.value,
        createdAt: fields._createdAt.value,
        status: fields._status.value,
        locale: fields._locale?.value,
        createdBy: fields._createdBy?.value,
        updatedBy: fields._updatedBy?.value,
        manageUrl: fields._manageUrl?.value,
        fields,
    };
}
export function lmdbStoredAssetToStackbit(storedAsset) {
    return {
        type: `asset`,
        id: storedAsset._objectId.value,
        _connectId: storedAsset.id,
        updatedAt: storedAsset._updatedAt.value,
        createdAt: storedAsset._createdAt.value,
        updatedBy: storedAsset._updatedBy?.value,
        createdBy: storedAsset._createdBy?.value,
        locale: storedAsset._locale?.value,
        manageUrl: storedAsset._manageUrl?.value,
        fields: {
            title: storedAsset.title,
            file: {
                type: `assetFile`,
                url: storedAsset.url.value,
                dimensions: {
                    width: storedAsset.width.value,
                    height: storedAsset.height.value,
                },
            },
        },
        status: `published`,
        context: {},
    };
}
